src/lib.rs

```diff
@@ -175,6 +181,68 @@ impl<'a> FMBuilder<'a> {
     }
 }
 
+/// An error indicating a failed match.
+/// The pattern and text are copied in so that the error isn't tied to their lifetimes.
+pub struct FMatchError {
+    ptn: String,
+    text: String,
+    fail_line: usize,
+}
+
+impl FMatchError {
+    pub fn failure_line(&self) -> usize {
+        self.fail_line
+    }
+}
+
+impl fmt::Display for FMatchError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // Figure out how many characters are required for the line numbers margin. Rust doesn't
+        // provide the numeric conversions we need, so we are using the `conv` crate here.
+        let max_lines = f64::value_from(usize::max(self.ptn.len(), self.text.len())).unwrap();
+        let lno_chars =
```

Comment:
I must admit, I find it easier to do something like `format!("{}", max_lines).len()`. It's less efficient, but that's not a worry here, and it's easier to read (and doesn't require an external library).

Reply:
I did consider this, but thought it was kind of silly to do a heap allocation just for this.

Reply:
I think it's probably better than compiling an extra crate, since this is the error path (i.e. we don't expect it to be actually taken).